docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,223 @@
+
+test('Assertion with Selector', async t => {
+    const developerNameInput = Selector('#developer-name');
+    const developerName      = Selector('#developer-name').value;
+
+    console.log(developerName);
+
+    await t
+        .wait(1000)
+        .expect(developerNameInput.value).eql('')
+        .typeText(developerNameInput, 'Peter')
+        .expect(developerName).eql('Peter');
+});
+```
+
+The `developerName` is now initialized with a re-executable Selector API promise. When a test controller receives this promise, it enables the smart assertion query mechanism and TestCafe can wait for the value to update.
+
+> The smart assertion query mechanism works with promises generated by TestCafe Selector API. Use the `await` keyword with user-created promises and promises generated by third-party libraries.
+
+This rule applies to ClientFunctions in the same capacity.
+
+The test below doesn't yield consistent results:
+
+```js
+import { Selector, ClientFunction } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with ClientFunction', async t => {
+    const getValue        = ClientFunction(() => document.getElementById('preferred-interface').value)
+    const interfaceSelect = Selector('#preferred-interface');
+    const interfaceOption = interfaceSelect.find('option');
+    const value           = await getValue();
+
+
+    await t
+        .click(interfaceSelect)
+        .click(interfaceOption.withText('JavaScript API'))
+        .expect(value).eql('JavaScript API')
+});
+```
+
+In this example, the client function obtains the value of a drop-down element. This value is then passed to the `t.expect` method. Since the client function is awaited when it's assigned to the variable, the value gets resolved instantly and never updates. The smart assertion query mechanism doesn't apply and the test fails.
+
+To solve the issue, pass the client function without the `await` keyword:
+
+```js
+import { Selector, ClientFunction } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with ClientFunction', async t => {
+    const getValue = ClientFunction(() => document.getElementById('preferred-interface').value)
+
+    const interfaceSelect = Selector('#preferred-interface');
+    const interfaceOption = interfaceSelect.find('option');
+    const value           =  getValue();
+
+    await t
+        .click(interfaceSelect)
+        .click(interfaceOption.withText('JavaScript API'))
+        .expect(value).eql('JavaScript API')
+});
+```
+
+In this case, TestCafe applies the smart query mechanism and retries the assertion multiple times. This test responds to the changes that happen on the page and is more conclusive.
+
+## Use of Page Objects
+
+Use the Page Model in your tests. It helps you abstract the tested page into a single file that you can use to target page elements.
```

Comment:
Do you have a feeling that when you first say what to use, and then explain what it is, resembles commercials?

Reply:
I'll introduce a code snippet and restructure this.